### PR TITLE
Improve workflow cache metrics

### PR DESCRIPTION
**Pull request by prathyushpv**
## What changed?
Add namespace ID tag to workflow cache metrics.
Add a new metric to track how long workflow lock is held.

## Why?
To get more insights of workflow cache based on namespace.

## How did you test it?
Unit test

## Potential risks
None

## Is hotfix candidate?
No

### CODE REVIEW

File: service/history/workflow/cache/cache.go [hunk 1]
```diff
@@ -332,16 +334,16 @@ func (c *cacheImpl) getOrCreateWorkflowExecutionInternal(
...
 	if err := c.lockWorkflowExecution(ctx, workflowCtx, cacheKey, lockPriority); err != nil {
 		metrics.CacheFailures.With(handler).Record(1)
 		metrics.AcquireLockFailedCounter.With(handler).Record(1)
 		return nil, nil, err
 	}
 
+	// TODO This will create a closure on every request.
+	//  Consider revisiting this if it causes too much GC activity
+	releaseFunc := c.makeReleaseFunc(cacheKey, shardContext, workflowCtx, forceClearContext, handler, time.Now())
+
```
Comment (stephanos) [inline service/history/workflow/cache/cache.go:346]: Nice drive-by improvement! 🙌

**Review by yycptt — APPROVED** [review]
I'd avoid namespaceID as the tag if possible, but I don't have very strong opinion.
The extra lookup could be a concern here, need to do some testing to tell.

**Comment by prathyushpv** [comment]
> I'd avoid namespaceID as the tag if possible, but I don't have very strong opinion. The extra lookup could be a concern here, need to do some testing to tell.

yeah. This is in the critical path of workflow operations. But it could be fine if it is cached in memory.

File: service/history/events/cache.go [hunk 2]
```diff
@@ -132,7 +132,7 @@ func (e *CacheImpl) validateKey(key EventKey) bool {
 }
 
 func (e *CacheImpl) GetEvent(ctx context.Context, shardID int32, key EventKey, firstEventID int64, branchToken []byte) (*historypb.HistoryEvent, error) {
-	handler := e.metricsHandler.WithTags(metrics.OperationTag(metrics.EventsCacheGetEventScope))
+	handler := e.metricsHandler.WithTags(metrics.OperationTag(metrics.EventsCacheGetEventScope), metrics.NamespaceTag(key.NamespaceID.String()))
```
Comment (yycptt) [inline service/history/events/cache.go:135]: Let's define a new tag for namespaceID then?

Comment (prathyushpv) [reply service/history/events/cache.go:135]: Done
